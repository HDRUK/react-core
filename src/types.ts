export type CommonVariants = "primary" | "secondary" | "tertiary";

export type CommonSizes = "small" | "medium" | "large";

export interface CommonStyleProps {
  display?: string;
  alignItems?: string;
  justifyContent?: string;
  p?: number;
  pl?: number;
  pb?: number;
  pt?: number;
  pr?: number;
  m?: number;
  ml?: number;
  mb?: number;
  mt?: number;
  mr?: number;
  flexGrow?: string;
  position?: string;
  width?: string;
  minWidth?: string;
  maxWidth?: string;
  height?: string;
  minHeight?: string;
  maxHeight?: string;
}

export type ComponentStyleProps = MarginStyleProps & WidthStyleProps;

export interface MarginStyleProps {
  m?: number;
  ml?: number;
  mb?: number;
  mt?: number;
  mr?: number;
}

export interface WidthStyleProps {
  width?: string;
  minWidth?: string;
  maxWidth?: string;
}

export interface HeightStyleProps {
  height?: string;
  minHeight?: string;
  maxHeight?: string;
}

export interface ThemeFontSizes {
  xxs: string;
  xs: string;
  sm: string;
  md: string;
  lg: string;
  xl: string;
  "2xl": string;
  "3xl": string;
  "4xl": string;
  "5xl": string;
}

export interface ThemeColors {
  [key: string]: string;
}

export interface ThemeComponents {
  [key: string]: string;
}

export interface Theme {
  base: {
    increment: number;
    unit: string;
  };
  font: {
    size: ThemeFontSizes;
  };
  colors: ThemeColors;
  components: ThemeComponents;
}
