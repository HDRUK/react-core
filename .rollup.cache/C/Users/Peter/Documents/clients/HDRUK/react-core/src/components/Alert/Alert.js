import { __assign, __rest } from "tslib";
import { jsx as _jsx, jsxs as _jsxs } from "@emotion/react/jsx-runtime";
/** @jsxImportSource @emotion/react */
import { useState, useCallback, useEffect } from "react";
import { ReactComponent as CloseIcon } from "../../images/icons/close.svg";
import { ReactComponent as DangerIcon } from "../../images/icons/danger.svg";
import { ReactComponent as InfoIcon } from "../../images/icons/info.svg";
import { ReactComponent as CheckIcon } from "../../images/icons/check.svg";
import Icon from "../Icon";
import { cx } from "@emotion/css";
import useCommonStyles from "../../hooks/useCommonStyles";
import * as styles from "./Alert.styles";
var Alert = function (_a) {
    var icon = _a.icon, variant = _a.variant, onClose = _a.onClose, children = _a.children, className = _a.className, mt = _a.mt, mb = _a.mb, ml = _a.ml, mr = _a.mr, m = _a.m, mx = _a.mx, my = _a.my, width = _a.width, minWidth = _a.minWidth, maxWidth = _a.maxWidth, dismissable = _a.dismissable, autoclose = _a.autoclose, autocloseDuration = _a.autocloseDuration, outerProps = __rest(_a, ["icon", "variant", "onClose", "children", "className", "mt", "mb", "ml", "mr", "m", "mx", "my", "width", "minWidth", "maxWidth", "dismissable", "autoclose", "autocloseDuration"]);
    var commonStyles = useCommonStyles({
        mt: mt,
        mb: mb,
        ml: ml,
        mr: mr,
        m: m,
        mx: mx,
        my: my,
        width: width,
        minWidth: minWidth,
        maxWidth: maxWidth,
    });
    var _b = useState(true), show = _b[0], setShow = _b[1];
    var handleClose = useCallback(function () {
        setShow(false);
        if (onClose)
            onClose();
    }, []);
    useEffect(function () {
        if (show && autoclose) {
            var showTimeout_1 = setTimeout(function () {
                handleClose();
            }, autocloseDuration);
            return function () {
                clearTimeout(showTimeout_1);
            };
        }
    }, [show, autoclose]);
    return show ? (_jsxs("div", __assign({ className: cx(className, "ui-Alert"), css: [commonStyles, styles.root({ variant: variant })] }, outerProps, { children: [_jsxs("div", __assign({ css: styles.icon }, { children: [icon, !icon && variant === "success" && (_jsx(Icon, { svg: _jsx(CheckIcon, { fill: "inherit" }), size: "xl" })), !icon && variant === "danger" && (_jsx(Icon, { svg: _jsx(DangerIcon, { fill: "inherit" }), size: "xl" })), !icon && variant === "warning" && (_jsx(Icon, { svg: _jsx(DangerIcon, { fill: "inherit" }), size: "xl" })), !icon && variant === "info" && (_jsx(Icon, { svg: _jsx(InfoIcon, { fill: "inherit" }), size: "xl" }))] })), _jsx("div", __assign({ css: styles.content }, { children: children })), dismissable && (_jsx("div", __assign({ css: styles.dismiss }, { children: _jsx(Icon, { svg: _jsx(CloseIcon, { fill: "inherit" }), onClick: handleClose, role: "button", size: "lg" }) })))] }))) : null;
};
Alert.defaultProps = {
    variant: "info",
    autocloseDuration: 10000,
};
export default Alert;
//# sourceMappingURL=Alert.js.map