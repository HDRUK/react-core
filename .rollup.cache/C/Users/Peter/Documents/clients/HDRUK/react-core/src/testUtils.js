// @ts-nocheck
import { __assign } from "tslib";
import React from "react";
import { ThemeProvider } from "@emotion/react";
import { render } from "@testing-library/react";
import { DEFAULT_THEME } from "./theme";
export var Providers = function (_a) {
    var children = _a.children;
    return React.createElement(ThemeProvider, { theme: DEFAULT_THEME }, children);
};
export var renderWithProviders = function (ui, options) {
    return render(ui, __assign({ wrapper: Providers }, options));
};
//# sourceMappingURL=testUtils.js.map