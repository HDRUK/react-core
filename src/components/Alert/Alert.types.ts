import { ComponentProps, ReactNode } from "react";
import { ComponentStyleProps } from "../../types";

export type AlertTypes = "success" | "warning" | "info" | "danger";

export interface AlertProps extends ComponentProps<"div">, ComponentStyleProps {
  variant: AlertTypes;
  icon?: ReactNode;
  autoclose?: boolean;
  autocloseDuration?: number;
  dismissable?: boolean;
  onClose?(): void;
}

export type AlertStyleProps = Pick<AlertProps, "variant">;
