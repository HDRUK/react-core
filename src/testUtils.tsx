import React from "react";
import { ThemeProvider } from "@emotion/react";
import { render, RenderOptions } from "@testing-library/react";
import { DEFAULT_THEME } from "./theme";

export const Providers: React.FC<{ children: React.ReactNode }> = ({
  children,
}) => {
  return <ThemeProvider theme={DEFAULT_THEME}>{children}</ThemeProvider>;
};

export const renderWithProviders = (
  ui: React.ReactElement,
  options?: RenderOptions
) => {
  return render(ui, { wrapper: Providers, ...options });
};
