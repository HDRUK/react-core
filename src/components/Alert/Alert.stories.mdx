import { useState } from "react";
import { Meta, Story, Canvas, ArgsTable } from "@storybook/addon-docs";
import { Card, CardBody } from "../Card";

import Alert from "./Alert";

<Meta title="Molecules/Alert" component={Alert} />

<Canvas>
    <Story name="Variants">
        {props => {
            return (
                <Card>
                    <CardBody>
                        {[
                            "success",
                            "info",
                            "warning",
                            "danger",
                            "description",
                        ].map(variant => (
                            <div key={variant}>
                                <Alert variant={variant} mb={4} dismissable>
                                    Lorem ipsum dolor sit amet ad hicio hortus
                                    est
                                </Alert>
                            </div>
                        ))}
                    </CardBody>
                </Card>
            );
        }}
    </Story>
    <Story name="Duration">
        {props => {
            const [showClosed, setShowClosed] = React.useState(false);
            const handleClose = React.useCallback(() => {
                setShowClosed(true);
            }, []);
            return (
                <Card>
                    <CardBody>
                        <Alert
                            variant="success"
                            autoclose
                            autocloseDuration={2000}
                            onClose={handleClose}>
                            Lorem ipsum dolor sit amet ad hicio hortus est
                        </Alert>
                        {showClosed && "Alert has been auto-closed!"}
                    </CardBody>
                </Card>
            );
        }}
    </Story>
</Canvas>

<ArgsTable of={Alert} />
