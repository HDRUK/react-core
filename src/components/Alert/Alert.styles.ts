import { css, Theme } from "@emotion/react";
import {
  getComponentGlobalStyles,
  getComponentVariantStyles,
  getSize,
} from "../../theme";
import { AlertStyleProps } from "./Alert.types";

export const root =
  ({ variant }: AlertStyleProps) =>
  (theme: Theme) => {
    return css`
      border-width: 2px;
      border-style: solid;
      padding: ${getSize(3, theme)};
      display: flex;

      ${getComponentGlobalStyles("Alert", theme)}
      ${getComponentVariantStyles("Alert", variant, theme)}
    `;
  };

export const icon = (theme: Theme) => css`
  margin-top: 0;
  margin-right: ${getSize(2, theme)};
  position: relative;
  top: 1px;
`;

export const dismiss = () => css`
  position: relative;
  top: 3px;
  cursor: pointer;
`;

export const content = () => css`
  flex-grow: 1;
  display: flex;
  align-items: center;

  > *:last-child {
    margin-bottom: 0;
  }
`;
