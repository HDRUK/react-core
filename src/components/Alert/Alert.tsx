/** @jsxImportSource @emotion/react */
import { useState, useCallback, useEffect } from "react";
import { ReactComponent as CloseIcon } from "../../images/icons/close.svg";
import { ReactComponent as DangerIcon } from "../../images/icons/danger.svg";
import { ReactComponent as InfoIcon } from "../../images/icons/info.svg";
import { ReactComponent as CheckIcon } from "../../images/icons/check.svg";
import Icon from "../Icon";

import { cx } from "@emotion/css";
import useCommonStyles from "../../hooks/useCommonStyles";
import * as styles from "./Alert.styles";
import { AlertProps } from "./Alert.types";

const Alert = ({
  icon,
  variant,
  onClose,
  children,
  className,
  mt,
  mb,
  ml,
  mr,
  m,
  mx,
  my,
  width,
  minWidth,
  maxWidth,
  dismissable,
  autoclose,
  autocloseDuration,
  ...outerProps
}: AlertProps) => {
  const commonStyles = useCommonStyles({
    mt,
    mb,
    ml,
    mr,
    m,
    mx,
    my,
    width,
    minWidth,
    maxWidth,
  });

  const [show, setShow] = useState(true);

  const handleClose = useCallback(() => {
    setShow(false);

    if (onClose) onClose();
  }, []);

  useEffect(() => {
    if (show && autoclose) {
      const showTimeout = setTimeout(() => {
        handleClose();
      }, autocloseDuration);

      return () => {
        clearTimeout(showTimeout);
      };
    }
  }, [show, autoclose]);

  return show ? (
    <div
      className={cx(className, "ui-Alert")}
      css={[commonStyles, styles.root({ variant })]}
      {...outerProps}
    >
      <div css={styles.icon}>
        {icon}
        {!icon && variant === "success" && (
          <Icon svg={<CheckIcon fill="inherit" />} size="xl" />
        )}
        {!icon && variant === "danger" && (
          <Icon svg={<DangerIcon fill="inherit" />} size="xl" />
        )}
        {!icon && variant === "warning" && (
          <Icon svg={<DangerIcon fill="inherit" />} size="xl" />
        )}
        {!icon && variant === "info" && (
          <Icon svg={<InfoIcon fill="inherit" />} size="xl" />
        )}
      </div>
      <div css={styles.content}>{children}</div>
      {dismissable && (
        <div css={styles.dismiss}>
          <Icon
            svg={<CloseIcon fill="inherit" />}
            onClick={handleClose}
            role="button"
            size="lg"
          />
        </div>
      )}
    </div>
  ) : null;
};

Alert.defaultProps = {
  variant: "info",
  autocloseDuration: 10000,
};

export default Alert;
