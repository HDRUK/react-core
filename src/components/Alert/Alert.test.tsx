import { fireEvent, RenderResult, waitFor } from "@testing-library/react";
import React from "react";
import { renderWithProviders } from "../../testUtils";
import Alert from "./Alert";
import { AlertProps } from "./Alert.types";

const props: AlertProps = {
  onClose: jest.fn(),
  variant: "success",
};

let wrapper: RenderResult;

describe("Given the Alert component", () => {
  describe("When it is rendered", () => {
    beforeEach(() => {
      wrapper = renderWithProviders(<Alert {...props}>Content</Alert>);
    });

    it("Then should match the snapshot", () => {
      expect(wrapper.asFragment()).toMatchSnapshot();
    });

    describe("And it is dismissable", () => {
      beforeEach(() => {
        wrapper.rerender(
          <Alert {...props} dismissable>
            Content
          </Alert>
        );
      });

      afterEach(() => {
        jest.clearAllMocks();
      });

      describe("And it is closed", () => {
        beforeEach(() => {
          const close = wrapper.getByRole("button");

          fireEvent.click(close);
        });

        it("Then does not show anything", async () => {
          await waitFor(() => {
            expect(wrapper.queryByText("Content")).toBeNull();
          });
        });

        it("Then calls onClose", () => {
          expect(props.onClose).toHaveBeenCalled();
        });
      });
    });

    describe("And autoclose is set", () => {
      beforeEach(() => {
        wrapper.rerender(
          <Alert {...props} autoclose autocloseDuration={1000}>
            Content
          </Alert>
        );
      });

      afterAll(() => {
        jest.clearAllMocks();
      });

      it("Then does not show anything", async () => {
        await waitFor(() => expect(wrapper.queryByText("Content")).toBeNull());
      });

      it("Then calls onClose", () => {
        expect(props.onClose).toHaveBeenCalled();
      });
    });

    describe("And it has an icon specified", () => {
      beforeEach(() => {
        wrapper.rerender(
          <Alert {...props} icon="Icon">
            Content
          </Alert>
        );
      });

      it("Then does not show other icons", async () => {
        await waitFor(() => expect(wrapper.getByText("Icon")).toBeTruthy());
      });
    });
  });
});
