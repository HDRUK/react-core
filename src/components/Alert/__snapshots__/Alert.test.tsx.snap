// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given the Alert component When it is rendered And it has an icon specified Then should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="ui-Alert css-zbu0kw-Alert"
  >
    <div
      class="css-183m98j-Alert"
    >
      Icon
    </div>
    <div
      class="css-1wsnpfb-Alert"
    >
      Content
    </div>
  </div>
</DocumentFragment>
`;

exports[`Given the Alert component When it is rendered Then should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="ui-Alert css-zbu0kw-Alert"
  >
    <div
      class="css-183m98j-Alert"
    >
      <span
        class="ui-Icon css-zkjjzt-Icon"
      >
        <reactcomponent
          fill="inherit"
        />
      </span>
    </div>
    <div
      class="css-1wsnpfb-Alert"
    >
      Content
    </div>
  </div>
</DocumentFragment>
`;
